src/Microsoft.AspNetCore.Blazor/Rendering/Renderer.cs

```diff
@@ -78,8 +78,13 @@ protected void DispatchEvent(int componentId, int eventHandlerId, UIEventArgs ev
                 // all in a single batch.
                 try
                 {
+                    if (!(handler is UIEventHandler uiEventHandler))
+                    {
+                        uiEventHandler = (UIEventArgs e) => handler.DynamicInvoke(e);
+                    }
+
                     _isBatchInProgress = true;
-                    GetRequiredComponentState(componentId).DispatchEvent(handler, eventArgs);
+                    GetRequiredComponentState(componentId).DispatchEvent(uiEventHandler, eventArgs);
```

Comment:
This was the trick. We only generate a 'wrapper' after we've processed the diff and decided to register the delegate for events. This allows the diff phase to leverage the delegate equality rules.